### PR TITLE
Replace LCA with naive implementations

### PR DESCRIPTION
**Pull request by rossbar**
I did a bit of investigation to see what it would take to solve #5869. The first step is to see what fails in the test suite when the existing LCA functions are replaced with the corresponding naive implementations. I'll break down the findings below.

Upon replacing the existing implementations with the new ones (see 50dc11f) we get 7 test failures from the existing LCA test suite:

<details>
  <summary>Initial test failures</summary>
  <pre>
networkx/algorithms/tests/test_lowest_common_ancestors.py:290: Failed           
=========================== short test summary info ============================
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor1
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor2
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor3
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor4
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor5
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor7
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor8
========================= 7 failed, 26 passed in 0.15s =========================
</pre>
</details>

These failures basically break down into two categories:
 1. Failing to raise exceptions
 2. "Wrong" LCA results (at least according to the test suite).

Let's address group 1. first.

The main problem here is that the `naive_all_pairs_lowest_common_ancestor` is a pure generator, whereas `all_pairs_lowest_common_ancestor` is actually a function that internally creates and returns a generator. Thus the input validation exceptions (i.e. checking it's a DAG, etc.) raise immediately in the case of `all_pairs_lowest_common_ancestor`, but aren't raised until the generator is *called* in the `naive_` case.

The simplest way to recover the original behavior is to copy the pattern used in `all_pairs_lowest_common_ancestor`; i.e. move the input validation checks *out* of the generator, create the generator *internally*, then return the generator instance - see 2c56aa7. Personally, I don't love this pattern (notice the repr has `<locals>` in it because of the internal definition, which is an implementation detail that users might find confusing if they look too closely), but it seems to be the easiest way to recover the prompt-exception behavior. Maybe there's a better way (`generator.throw?`)?

Anyways 2c56aa7 gets rid of *almost all* of the exception-raising test failures:

<details>
  <summary>After internal generator fix</summary>
  <pre>
============================= test session starts ==============================
platform linux -- Python 3.10.5, pytest-7.1.1, pluggy-1.0.0
Matplotlib: 3.5.1
Freetype: 2.6.1
rootdir: /home/ross/repos/networkx
plugins: mpl-0.14.0, cov-3.0.0
collected 33 items

networkx/algorithms/tests/test_lowest_common_ancestors.py .............F [ 42%]
FFFF...............                                                      [100%]

=================================== FAILURES ===================================
______________ TestDAGLCA.test_all_pairs_lowest_common_ancestor1 _______________

self = <networkx.algorithms.tests.test_lowest_common_ancestors.TestDAGLCA object at 0x7f3e5c650490>

    def test_all_pairs_lowest_common_ancestor1(self):
        """Produces the correct results."""
>       self.assert_lca_dicts_same(dict(all_pairs_lca(self.DG)), self.gold)

networkx/algorithms/tests/test_lowest_common_ancestors.py:239: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <networkx.algorithms.tests.test_lowest_common_ancestors.TestDAGLCA object at 0x7f3e5c650490>
d1 = {(0, 0): 0, (0, 1): 0, (0, 2): 0, (0, 3): 0, ...}
d2 = {(0, 0): 0, (0, 1): 0, (0, 2): 0, (0, 3): 0, ...}
G = <networkx.classes.digraph.DiGraph object at 0x7f3e5c6536d0>

    def assert_lca_dicts_same(self, d1, d2, G=None):
        """Checks if d1 and d2 contain the same pairs and
        have a node at the same distance from root for each.
        If G is None use self.DG."""
        if G is None:
            G = self.DG
            root_distance = self.root_distance
        else:
            roots = [n for n, deg in G.in_degree if deg == 0]
            assert len(roots) == 1
            root_distance = nx.shortest_path_length(G, source=roots[0])
    
        for a, b in ((min(pair), max(pair)) for pair in chain(d1, d2)):
>           assert (
                root_distance[get_pair(d1, a, b)] == root_distance[get_pair(d2, a, b)]
            )
E           assert 2 == 3

networkx/algorithms/tests/test_lowest_common_ancestors.py:233: AssertionError
______________ TestDAGLCA.test_all_pairs_lowest_common_ancestor2 _______________

self = <networkx.algorithms.tests.test_lowest_common_ancestors.TestDAGLCA object at 0x7f3e5c650670>

    def test_all_pairs_lowest_common_ancestor2(self):
        """Produces the correct results when all pairs given."""
        all_pairs = list(product(self.DG.nodes(), self.DG.nodes()))
        ans = all_pairs_lca(self.DG, pairs=all_pairs)
>       self.assert_lca_dicts_same(dict(ans), self.gold)

networkx/algorithms/tests/test_lowest_common_ancestors.py:245: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <networkx.algorithms.tests.test_lowest_common_ancestors.TestDAGLCA object at 0x7f3e5c650670>
d1 = {(0, 0): 0, (0, 1): 0, (0, 2): 0, (0, 3): 0, ...}
d2 = {(0, 0): 0, (0, 1): 0, (0, 2): 0, (0, 3): 0, ...}
G = <networkx.classes.digraph.DiGraph object at 0x7f3e5c6536d0>

    def assert_lca_dicts_same(self, d1, d2, G=None):
        """Checks if d1 and d2 contain the same pairs and
        have a node at the same distance from root for each.
        If G is None use self.DG."""
        if G is None:
            G = self.DG
            root_distance = self.root_distance
        else:
            roots = [n for n, deg in G.in_degree if deg == 0]
            assert len(roots) == 1
            root_distance = nx.shortest_path_length(G, source=roots[0])
    
        for a, b in ((min(pair), max(pair)) for pair in chain(d1, d2)):
>           assert (
                root_distance[get_pair(d1, a, b)] == root_distance[get_pair(d2, a, b)]
            )
E           assert 2 == 3

networkx/algorithms/tests/test_lowest_common_ancestors.py:233: AssertionError
______________ TestDAGLCA.test_all_pairs_lowest_common_ancestor3 _______________

self = <networkx.algorithms.tests.test_lowest_common_ancestors.TestDAGLCA object at 0x7f3e5c650850>

    def test_all_pairs_lowest_common_ancestor3(self):
        """Produces the correct results when all pairs given as a generator."""
        all_pairs = product(self.DG.nodes(), self.DG.nodes())
        ans = all_pairs_lca(self.DG, pairs=all_pairs)
>       self.assert_lca_dicts_same(dict(ans), self.gold)

networkx/algorithms/tests/test_lowest_common_ancestors.py:251: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <networkx.algorithms.tests.test_lowest_common_ancestors.TestDAGLCA object at 0x7f3e5c650850>
d1 = {(0, 0): 0, (0, 1): 0, (0, 2): 0, (0, 3): 0, ...}
d2 = {(0, 0): 0, (0, 1): 0, (0, 2): 0, (0, 3): 0, ...}
G = <networkx.classes.digraph.DiGraph object at 0x7f3e5c6536d0>

    def assert_lca_dicts_same(self, d1, d2, G=None):
        """Checks if d1 and d2 contain the same pairs and
        have a node at the same distance from root for each.
        If G is None use self.DG."""
        if G is None:
            G = self.DG
            root_distance = self.root_distance
        else:
            roots = [n for n, deg in G.in_degree if deg == 0]
            assert len(roots) == 1
            root_distance = nx.shortest_path_length(G, source=roots[0])
    
        for a, b in ((min(pair), max(pair)) for pair in chain(d1, d2)):
>           assert (
                root_distance[get_pair(d1, a, b)] == root_distance[get_pair(d2, a, b)]
            )
E           assert 2 == 3

networkx/algorithms/tests/test_lowest_common_ancestors.py:233: AssertionError
______________ TestDAGLCA.test_all_pairs_lowest_common_ancestor4 _______________

self = <networkx.algorithms.tests.test_lowest_common_ancestors.TestDAGLCA object at 0x7f3e5c650a30>

    def test_all_pairs_lowest_common_ancestor4(self):
        """Graph with two roots."""
        G = self.DG.copy()
        G.add_edge(9, 10)
        G.add_edge(9, 4)
        gold = self.gold.copy()
        gold[9, 9] = 9
        gold[9, 10] = 9
        gold[9, 4] = 9
        gold[9, 3] = 9
        gold[10, 4] = 9
        gold[10, 3] = 9
        gold[10, 10] = 10
    
        testing = dict(all_pairs_lca(G))
    
        G.add_edge(-1, 9)
        G.add_edge(-1, 0)
>       self.assert_lca_dicts_same(testing, gold, G)

networkx/algorithms/tests/test_lowest_common_ancestors.py:271: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <networkx.algorithms.tests.test_lowest_common_ancestors.TestDAGLCA object at 0x7f3e5c650a30>
d1 = {(0, 0): 0, (0, 1): 0, (0, 2): 0, (0, 3): 0, ...}
d2 = {(0, 0): 0, (0, 1): 0, (0, 2): 0, (0, 3): 0, ...}
G = <networkx.classes.digraph.DiGraph object at 0x7f3e5c55b7f0>

    def assert_lca_dicts_same(self, d1, d2, G=None):
        """Checks if d1 and d2 contain the same pairs and
        have a node at the same distance from root for each.
        If G is None use self.DG."""
        if G is None:
            G = self.DG
            root_distance = self.root_distance
        else:
            roots = [n for n, deg in G.in_degree if deg == 0]
            assert len(roots) == 1
            root_distance = nx.shortest_path_length(G, source=roots[0])
    
        for a, b in ((min(pair), max(pair)) for pair in chain(d1, d2)):
>           assert (
                root_distance[get_pair(d1, a, b)] == root_distance[get_pair(d2, a, b)]
            )
E           assert 3 == 4

networkx/algorithms/tests/test_lowest_common_ancestors.py:233: AssertionError
______________ TestDAGLCA.test_all_pairs_lowest_common_ancestor5 _______________

self = <networkx.algorithms.tests.test_lowest_common_ancestors.TestDAGLCA object at 0x7f3e5c650c10>

    def test_all_pairs_lowest_common_ancestor5(self):
        """Test that pairs not in the graph raises error."""
>       pytest.raises(nx.NodeNotFound, all_pairs_lca, self.DG, [(-1, -1)])
E       Failed: DID NOT RAISE <class 'networkx.exception.NodeNotFound'>

networkx/algorithms/tests/test_lowest_common_ancestors.py:275: Failed
=========================== short test summary info ============================
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor1
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor2
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor3
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor4
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor5
========================= 5 failed, 28 passed in 0.13s =========================

</pre>
</details>

The next set of failures are due to the `root_distance` not being as expected. Fortunately, @dtekinoglu, @MridulS , and @dschult noticed that there was an error in one of the ground-truth values for one of the tests. If we fix that, then these errors disappear - see 3bf51ac.

<details>
  <summary>Test failures after fixing wrong self-ancestor in test reference</summary>
  <pre>
============================= test session starts ==============================
platform linux -- Python 3.10.5, pytest-7.1.1, pluggy-1.0.0
Matplotlib: 3.5.1
Freetype: 2.6.1
rootdir: /home/ross/repos/networkx
plugins: mpl-0.14.0, cov-3.0.0
collected 33 items

networkx/algorithms/tests/test_lowest_common_ancestors.py .............. [ 42%]
...F...............                                                      [100%]

=================================== FAILURES ===================================
______________ TestDAGLCA.test_all_pairs_lowest_common_ancestor5 _______________

self = <networkx.algorithms.tests.test_lowest_common_ancestors.TestDAGLCA object at 0x7f8d806f4b80>

    def test_all_pairs_lowest_common_ancestor5(self):
        """Test that pairs not in the graph raises error."""
>       pytest.raises(nx.NodeNotFound, all_pairs_lca, self.DG, [(-1, -1)])
E       Failed: DID NOT RAISE <class 'networkx.exception.NodeNotFound'>

networkx/algorithms/tests/test_lowest_common_ancestors.py:275: Failed
=========================== short test summary info ============================
FAILED networkx/algorithms/tests/test_lowest_common_ancestors.py::TestDAGLCA::test_all_pairs_lowest_common_ancestor5
========================= 1 failed, 32 passed in 0.10s =========================
</pre>
</details>

So that's good news... it looks like the original LCA implementation was only passing the test case because the test case itself was obviously incorrect! The "naive" implementation fails the incorrect test and passes the correct one - just one more indication that the naive implementations are more correct and reliable than the existing ones!

At this point there's one failure remaining, and it's another failure to raise a prompt exception. In this case, `all_pairs_lowest_common_ancestor` actually did pre-evaluation of the `pairs` argument to verify that all of the nodes provided in `pairs` are actually in `G`. In order to match this behavior, we'd have to add the same pre-processing to the naive implementations. Unfortunately, there's another quirk that makes this a bit clunkier: the docstring for `all_pairs_lowest_common_ancestor` says that `pairs` must be an *iterable* of 2-tuples of nodes, but the test suite also explicitly tests that *iterators* (e.g. generators) are accepted (see the `_ancestor3` test). This means that, in order to match the original behavior, we have to add an extra conversion step to accept both inputs (search for `pairset` in the original implementation). This is all done in b535ca1. Again, not the prettiest solution, but the easiest way (that I could think of at least) to recover the original behavior of `all_pairs_lowest_common_ancestor`.

In summary, a few changes would be required to replace the LCA function implementations. Having gone through this exercise, I think it's definitely worthwhile because one of those changes actually shows how the current implementation gives flat-out wrong results.

I'm going to open this as a draft PR and hope to get feedback on the individual changes made in each of the three commits 2c56aa7, 3bf51ac, and b535ca1. If those look acceptable, then I remove the aliasing in 50dc11f and replace the implementations of `all_pairs_lowest_common_ancestor` and `lowest_common_ancestor` with the updated implementations.

### CODE REVIEW

File: networkx/algorithms/lowest_common_ancestors.py [hunk 1]
```diff
@@ -72,34 +72,47 @@ def naive_all_pairs_lowest_common_ancestor(G, pairs=None):
...
-        if common_ancestors:
-            common_ancestor = next(iter(common_ancestors))
-            while True:
-                successor = None
-                for lower_ancestor in G.successors(common_ancestor):
-                    if lower_ancestor in common_ancestors:
-                        successor = lower_ancestor
+    else:
+        # Convert iterator to iterable, if necessary, trim duplicates
+        pairs = set(pairs)
```
Comment (rossbar) [inline networkx/algorithms/lowest_common_ancestors.py:79]: Using a set here can/will break the order in which the pairs+lca are yielded. If order is important, `list` can be used instead.

**Review by dschult — COMMENTED** [review]
Wow -- nice work here!  There are certainly a lot of parts to this.

I have one suggestion below based on your comment
And I think this is ready to replace the previous functions.

File: networkx/algorithms/lowest_common_ancestors.py (hunk 1 above)
Comment (rossbar) [reply networkx/algorithms/lowest_common_ancestors.py:79]: Good idea! Ultimately I think it's strange that the function guarantees *both* that it eagerly raises `NodeNotFound` exceptions for `pairs` *and* excepts generators for the `pairs` argument --- The former essentially guarantees that the generator either needs to be consumed and stored in some iterable under-the-hood (as is currently done, obviating the advantages of using a generator), or the generator needs to be cycled through twice.

However this is a discussion for another issue/PR! For now I'll focus on getting the behavior exactly the same with *as minimal* changes as possible.

Comment (dschult) [reply networkx/algorithms/lowest_common_ancestors.py]: I understand the strangeness about both eagerly checking for nodes and accepting a generator. But it is more friendly for the user not to have to construct e.g. a list if they have an iterator. And they don't have to decide whether it should be a list/dict/set/tuple. At the same time it is more friendly to raise NodeNotFound eagerly. What is the downside? Perhaps that there are too many different possible inputs, which makes the code harder to read/follow.

It certainly means that we are not able to maintain the generator processing flow. We convert it to a dict/set in order to test the nodes.  If we were designing from scratch I would be tempted to make the input ignore pairs which contain nodes not in G. Then we can eliminate the NodeNotFound checking code...  

I think there is probably a good general principle here -- but I'm not sure what it is :)
As you say -- maybe that's for another PR/issue.

**Comment by rossbar** [comment]
Given the positive feedback on the changes that had to be made for consistency (thanks @dschult !) I went ahead and did the replacement.

Just to recap: this replaces the existing implementations of `all_pairs_lowest_common_ancestor` and `lowest_common_ancestor` with the naive implementations from #5736 and removes the `naive_all_pairs_lowest_common_ancestor` and `naive_lowest_common_ancestor` functions. Removing the functions shouldn't require a deprecation or anything because they haven't been included in a release, though I did add a bullet to the release notes about the change in implementation to the existing LCA functions.

I also removed the *duplicate* tests (i.e. the tests that were the same for the `naive_` and non-`naive_` versions of the functions) but kept the new ones that were added in #5736 that highlight cases that used to fail with the old algorithm.

**Review by dschult — APPROVED** [review]
This looks good!  
I know @rossbar said we should refactor the commits slightly, and I don't know much about that process -- I think all the commits get squashed at some point anyway, so maybe that's a change to the message when we press the merge button... 

Anyway, I'm approving this -- I think it is ready!  Thanks @rossbar and especially @dtekinoglu

**Review by jarrodmillman — APPROVED** [review]
